test/vfd_swmr_group_writer.c

```diff
@@ -171,30 +244,1538 @@ state_init(state_t *s, int argc, char **argv)
+    
+    unsigned max_compact = 0;
+    unsigned min_dense = 0;
+    unsigned u = 0;
+
+    char attrname[VS_ATTR_NAME_LEN];
+    const char* aname_format="attr-%u-%u";
+    const char* adname_format="attr-d-%u-%u";
+
+    /* Obtain the maximal number of attributes to be stored in compact
+     * storage and the minimal number of attributes to be stored in
+     * dense storage. */
+    if (H5Pget_attr_phase_change(gcpl, &max_compact, &min_dense) < 0) {
+        H5_FAILED(); AT();
+        printf("H5Pget_attr_phase_change failed\n");
+        goto error;
+    }
+    u= max_compact +1;
+
+#if 0
+    if(max_compact < min_dense) 
+        printf("Minimum number of attributes stored in dense storage should be less than maximum number of attributes stored in compact storage.\n");
+#endif
+
+    // delete a number of attributes so that the attribute storage just becomes dense.
+    for(u--;u>=(min_dense-1);u--) {
+        HDsprintf(attrname, aname_format, which,max_compact-u);
+        if (H5Adelete(obj_id,attrname) < 0) {
+            H5_FAILED(); AT();
+            printf("H5Adelete failed\n");
+            goto error;
+        }
+
+        /* For each attribute deletion, we want to ensure the verification
+         * from the reader. 
+         * So writer sends a message to reader: an attribute is successfully deleted. 
+           then wait for reader to verify and send an acknowledgement message back. */
+        if(s->use_named_pipes && s->attr_test == true) {
+            dbgf(2, "writer: delete attr - ready to send the message: %d\n", s->np_notify+1);
+            if(np_wr_send_receive(s) == false) {
+                H5_FAILED(); AT();
+                dbgf(2, "writer: delete attr - verification failed.\n");
+                goto error2;
+            }
+        }
+    }
+
+    /* The writer deletes another attribute, the storage is
+     * still dense. However, the attribute to be deleted
+     * doesn't follow the previous for loop. It may be 
+     * in different location in the object header. Just add
+     * a litter variation to check if this operation is successful. 
+     * The attribute name to be deleted is attr-max_compact+which-0
+     */
+     
+    HDsprintf(attrname,adname_format,max_compact+which,0);
+    if (H5Adelete(obj_id,attrname) < 0) {
+            H5_FAILED(); AT();
+            printf("H5Adelete failed\n");
+            goto error;
+    }
+    /* Again we need to notify the reader via Named pipe. */
+    if(s->use_named_pipes && s->attr_test == true) {
+        dbgf(2, "writer: delete attr - ready to send the message: %d\n", s->np_notify+1);
+        if(np_wr_send_receive(s) == false) {
+            H5_FAILED(); AT();
+            dbgf(2, "writer: delete attr - verification failed.\n");
+            goto error2;
+        }
+    }
+
+    /* The following comments are left here in case in the future we want to
+     * use HDF5 function to verify the attribute storage */
```

Comment:
I am inclined to do the verification of storage now instead of later.

Reply:
This is a question I want to check how you did this. I don't find an obvious API and documentation how this can be done. In your code, I see the following block:` /* Get the object information */
    if(H5Oget_native_info(did, &ninfo, H5O_NATIVE_INFO_HDR|H5O_NATIVE_INFO_META_SIZE) < 0) {
        printf("H5Oget_native_info failed\n");
        TEST_ERROR;
    }

    if(!which) {
        dbgf(2, "Verifying no cont\n");
        return(ninfo.hdr.nchunks == 1);

    } else if(which < cut_point) {
        dbgf(2, "Verifying storage compact & cont\n");
        return(ninfo.meta_size.attr.index_size == 0 && 
               ninfo.meta_size.attr.heap_size == 0 &&
               ninfo.hdr.nchunks > 1);
    } else {
        dbgf(2, "Verifying storage dense & no cont\n");
        return(ninfo.meta_size.attr.index_size != 0 && 
               ninfo.meta_size.attr.heap_size != 0 &&
               ninfo.hdr.nchunks == 1);
    }
`

Reply:
ninfo.meta_size.attr.index_size  and ninfo.meta_size.attr.heap_size are metadata size for data structures used to track dense storage.  They which will be zero for compact storage but nonzero for dense storage.
ninfo.hdr.nchunks is the # of chunks for the object header.  This will be 1 if there is just one chunk but > 1 if there are continuation chunks.

Reply:
Thanks, that's what I think. I will use <ninfo.meta_size.attr.index_size and ninfo.meta_size.attr.heap_size>  to check the dense or the compact in the code. The <ninfo.hdr.nchunks> is irrelevant to the group attribute case. So I can ignore.